SQL-119: OIDC/SSO authentication documentation (#35591)
MaterializeInc/materialize: doc/user/content/security/self-managed/sso.md

To configure Self-Managed Materialize for OIDC authentication, update the
following fields:

| Resource | Configuration | Description
|----------|---------------| ------------
| Materialize CR | `spec.authenticatorKind` | Set to `Oidc` to enable OIDC authentication.
| Kubernetes Secret | `external_login_password_mz_system` | Specify the password for the `mz_system` user, who is needed for initial system parameter configuration. Add `external_login_password_mz_system` to the Kubernetes Secret referenced in the Materialize CR's `spec.backendSecretName` field.

The following example Kubernetes manifest includes configuration for OIDC
authentication:

```hc {hl_lines="15 25"}
apiVersion: v1
kind: Namespace
metadata:
  name: materialize-environment
---
apiVersion: v1
kind: Secret
metadata:
  name: materialize-backend
  namespace: materialize-environment
stringData:
  metadata_backend_url: "..."
  persist_backend_url: "..."
  license_key: "..."
  external_login_password_mz_system: "enter_mz_system_password"
---
apiVersion: materialize.cloud/v1alpha1
kind: Materialize
metadata:
  name: 12345678-1234-1234-1234-123456789012
  namespace: materialize-environment
spec:

>   environmentdImageRef: materialize/environmentd:vX.Y.Z

Author: Addressed here https://github.com/MaterializeInc/materialize/pull/35591/changes/821c27e10d1b0502cf47fae0af009a268665b8f2

`oidc_audience` in production environments.
{{</ warning >}}

### Configure via ConfigMap

Create a ConfigMap with your OIDC parameters and reference it in the Materialize
CR's `spec.systemParameterConfigmapName` field:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: mz-system-params
  namespace: materialize-environment
data:
  system-params.json: |
    {
      "oidc_issuer": "https://your-org.okta.com/oauth2/default",
      "oidc_audience": "[\"YOUR_CLIENT_ID\"]",
      "oidc_authentication_claim": "email",
      "console_oidc_client_id": "YOUR_CLIENT_ID",
      "console_oidc_scopes": "openid"
    }
```

Then add the ConfigMap name to your Materialize CR:

```yaml
spec:
  systemParameterConfigmapName: mz-system-params
```

### Configure via SQL

Alternatively, connect as `mz_system` and set the parameters using
`ALTER SYSTEM SET`:

```mzsql
ALTER SYSTEM SET oidc_issuer = 'https://your-org.okta.com/oauth2/default';
ALTER SYSTEM SET oidc_audience = '["YOUR_CLIENT_ID"]';
ALTER SYSTEM SET oidc_authentication_claim = 'email';
ALTER SYSTEM SET console_oidc_client_id = 'YOUR_CLIENT_ID';
ALTER SYSTEM SET console_oidc_scopes = 'openid';
```

## Step 4. Verify the configuration

> 1. Navigate to your Materialize console. You should see an option to sign in

Author: Addressed here https://github.com/MaterializeInc/materialize/pull/35591/changes/2df1c84cf00aea73be1cc0d136664893f5e05e76

{{< important >}}
The `oidc_auth_enabled=true` connection option is **required** for OIDC
authentication over the PostgreSQL wire protocol. Without it, Materialize
defaults to password authentication.
{{</ important >}}

### Get your token from the console

The easiest way to get an ID token is through the Materialize console:

1. Sign in to the console via SSO.
1. Click **Connect**.
1. Copy the provided ID token.

### Connect with psql

Use the ID token as your password:

```shell
PGOPTIONS='-c oidc_auth_enabled=true' \
PGPASSWORD="<your-id-token>" \
psql -h <materialize-host> -p 6875 -U <username> materialize
```

Replace `<username>` with the value of the authentication claim in your JWT
(e.g., your email address if `oidc_authentication_claim` is set to `email`).

{{< note >}}
Materialize validates the token at **connection time only**. Once a connection
is established, it persists until disconnected, regardless of token expiry.
{{</ note >}}

### Get a token using CLI tools

You can also obtain an ID token outside the console using OIDC CLI tools such as
[`oidc-agent`](https://indigo-dc.gitbook.io/oidc-agent) or
[`oauth2c`](https://github.com/cloudentity/oauth2c).

> ## Auto-provisioned roles

Reviewer: > The role name is derived from the JWT claim specified by oidc_authentication_claim (default: sub).
This felt confusing to me - I reckon we can remove it

1. In the Okta Admin Console, go to **Applications** > **Applications** and
   click **Create App Integration**.

1. Select **OIDC - OpenID Connect** as the sign-in method and **Single-Page
   Application** as the application type. Click **Next**.

1. Configure the application:
   - **App integration name**: Enter a name (e.g., `Materialize`).
   - **Grant type**: Ensure **Authorization Code** is selected (PKCE is used
     automatically for single-page applications).
   - **Sign-in redirect URIs**: Enter
     `https://<your-console-domain>/auth/callback`.
   - **Sign-out redirect URIs**: Optionally, enter
     `https://<your-console-domain>/account/login`.

1. Click **Save**.

1. On the application's **General** tab, note the **Client ID**.

   *Single-page applications use PKCE instead of a client secret. You do not
   need a client secret for the console application.*

1. Go to **Security** > **API** and note your **Issuer URI** from the
   authorization server you want to use (e.g.,
   `https://your-org.okta.com/oauth2/default`).

1. Go to the **Assignments** tab and assign the users or groups that should have

>    access to Materialize.

Reviewer 2: I'm wondering ... if this is where we're setting up users,  should blurb about if migrating, these need to match existing MZ databas roles and link to the migration page?  (In particular where we find existing login users?)

{{% include-headless
"/headless/self-managed-deployments/enabled-auth-setting-warning" %}}

## Step 3. Configure OIDC system parameters

Configure the OIDC system parameters to connect Materialize to your identity
provider. You can use either a
[ConfigMap](/self-managed-deployments/configuration-system-parameters/#configure-system-parameters-via-configmap)
or SQL commands, but it is strongly recommended to use a ConfigMap.

### OIDC system parameters

| Parameter | Description | Required | Default |
|-----------|-------------|----------|---------|
| `oidc_issuer` | The OIDC issuer URL (e.g., `https://your-org.okta.com/oauth2/default`). Materialize uses this to discover the JWKS endpoint for token validation. | Yes | None |
| `oidc_audience` | A JSON array of expected audience values for token validation (e.g., `["your-client-id"]`). Use the **client ID from [Step 1](#step-1-configure-your-identity-provider)**. Materialize checks that the JWT's `aud` claim contains at least one of these values. **By default, this is empty, and audience validation is skipped.**| No | `[]` |
| `oidc_authentication_claim` | The JWT claim to use as the Materialize username. | No | `sub` |

>  | `oidc_audience` | A JSON array of expected audience values for token validation (e.g., `["your-client-id"]`). Use the **client ID from [Step 1](#step-1-configure-your-identity-provider)**. Materialize checks that the JWT's `aud` claim contains at least one of these values. **By default, this is empty, and audience validation is skipped.** | No | `[]` |

Author: Resolved in [docs: Remove duplicate oidc_audience row in SSO parameters table](https://github.com/MaterializeInc/materialize/pull/35591/changes/7db9d598a684b7199dcf861f91ed01857ffe75c5)

CR's `spec.systemParameterConfigmapName` field. At this point, your manifest should look like:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: mz-system-params
  namespace: materialize-environment
data:
  system-params.json: |
    {
      "oidc_issuer": "YOUR_OIDC_ISSUER",
      "oidc_audience": "[\"YOUR_CLIENT_ID\"]",
      "oidc_authentication_claim": "email",
      "console_oidc_client_id": "YOUR_CLIENT_ID",
      "console_oidc_scopes": "openid email"
    }
---
apiVersion: v1
kind: Secret
metadata:
  name: materialize-backend
  namespace: materialize-environment
stringData:
  metadata_backend_url: ...
  persist_backend_url: ...
  license_key: ...
  external_login_password_mz_system: "enter_mz_system_password"
---
apiVersion: materialize.cloud/v1alpha1
kind: Materialize
metadata:
  name: 12345678-1234-1234-1234-123456789012
  namespace: materialize-environment
spec:
  authenticatorKind: Oidc
  backendSecretName: materialize-backend
  requestRollout: 00000000-0000-0000-0000-000000000003 # Switching to Oidc requires a rollout
  systemParameterConfigmapName: mz-system-params
```

### Configure via SQL

> Alternatively, connect as `mz_system` and set the parameters using

Author: Resolved in [docs: Note mz_system always uses password auth in SQL config section](https://github.com/MaterializeInc/materialize/pull/35591/changes/587dc0589b828cd282a5b40b0c7c9c3bd7aeece8)

> Alternatively, connect as `mz_system` and set the parameters using

Reviewer 2: So ... Are you sure if the oidc system params aren't set up prior to this step that you can login as mz_system?

`ALTER SYSTEM SET`:

```mzsql
ALTER SYSTEM SET oidc_issuer = 'https://your-org.okta.com/oauth2/default';
ALTER SYSTEM SET oidc_audience = '["YOUR_CLIENT_ID"]';
ALTER SYSTEM SET oidc_authentication_claim = 'email';
ALTER SYSTEM SET console_oidc_client_id = 'YOUR_CLIENT_ID';
ALTER SYSTEM SET console_oidc_scopes = 'openid';
```

## Step 4. Verify the configuration

1. Navigate to your Materialize console. You should see an option to sign in
   via your identity provider via "Use single sign-on".

   ![Materialize console login screen showing the SSO sign-in
   option](/images/console/console-self-managed-sso.png "Materialize console login screen
   with SSO option")

1. Sign in through your IdP. After successful authentication, you are redirected
   back to the Materialize console.

> 1. To confirm which SQL role you've signed in as via SSO, open the [SQL Shell](/console/sql-shell/) in the Materialize console. In the welcome message, you should see the role's name labelled under "User". This is derived from the `oidc_authentication_claim` claim in your identity token:

Reviewer 2: Maybe not "SQL role"

---
title: "Single sign-on (SSO)"
description: "Configure OIDC-based single sign-on (SSO) for Self-Managed Materialize."
menu:
  main:
    parent: "security-sm"
    name: "Single sign-on (SSO)"
    identifier: "sso-sm"
    weight: 9

> ---

Reviewer 3: **Dismissed candidates (informational, no action requested):** [`rule_065`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_065.yaml) — *add flattened section alias when page lives in a subsection*, confidence 5 — and [`rule_066`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_066.yaml) — *add self-managed versioned redirect alias*, confidence 5.

Both rules' triggers fire on this new page's frontmatter (the file lives under `/security/self-managed/sso/` so a flattened `/security/sso/` alias is mechanically applicable, and self-managed pages typically carry versioned redirect aliases).

**Why dismissed:**
- These rules' corpus examples are restructuring moves where an *old* URL must keep resolving. This is a **new** page — no prior URL to redirect from, so the alias has nothing to redirect.
- The skill's `frontmatter` constraint forbids unilateral YAML changes without explicit user instruction.

If there's an old cloud SSO URL (`/manage/access-control/sso/` or similar) you intend to redirect to this new self-managed page, adding the `aliases:` entry would be appropriate — that's an editorial decision for you.

{{< public-preview />}}

Single sign-on (SSO) allows users to authenticate to Self-Managed Materialize
using their organization's identity provider (IdP) via
[OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/).
Instead of managing passwords directly in Materialize, users sign in through
their IdP (e.g., Okta, Microsoft Entra ID) and receive a JWT token that
Materialize validates.

> {{< note >}}
> SSO handles **authentication** only. Permissions within the database are managed
> separately using [role-based access control (RBAC)](/security/self-managed/access-control/).
> {{</ note >}}

Reviewer 3: [`rule_043`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_043.yaml) — *normalize blank-line padding inside admonition blocks*, confidence 5.

```suggestion
{{< note >}}

SSO handles **authentication** only. Permissions within the database are managed
separately using [role-based access control (RBAC)](/security/self-managed/access-control/).

{{</ note >}}
```

> ## Before you begin

Reviewer 3: [`rule_037`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_037.yaml) — *rename before-you-begin heading to prerequisite*, confidence 4. House style is `## Prerequisites` (plural here — two prereqs listed).

```suggestion
## Prerequisites
```

Make sure you have:

- An OIDC-capable identity provider (e.g., Okta, Microsoft Entra ID, or any
  provider that supports OpenID Connect).
- Admin access to your Kubernetes cluster where Materialize is deployed.

> ## Step 1. Configure your identity provider

Reviewer 3: **Dismissed candidate (informational, no action requested):** [`rule_057`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_057.yaml) — *demote step headings and drop step prefix*, confidence 4.

The rule's trigger matches all four `## Step N.` headings on this page (L34, L180, L229, L332). It would demote them to `### N.` under a newly-introduced parent H2.

**Why dismissed:**
- The rule's own `do_not_apply_when` covers this case: *"the step is a standalone top-level page section that has no logical parent H2 to nest under."* These `## Step N.` headings are the page's primary structure — there's no parent section to nest them under.
- The skill's `heading_hierarchy` constraint forbids silent heading-level changes without explicit user approval.

Surfaced for transparency in case you want to restructure (e.g., introduce a `## Setup` parent H2). Otherwise, leave as-is.

> ## Step 1. Configure your identity provider

Reviewer 3: **Dismissed candidate (informational, no action requested):** [`rule_007`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_007.yaml) — *normalize procedural headings to lettered sections with numbered subsections*, confidence 3.

An alternative restructure to rule_057. Would re-group the four step headings into lettered top-level sections, e.g.:
- `## A. Configure your identity provider` (current Step 1)
- `## B. Configure Materialize` (current Steps 2 + 3, both configure the Materialize deployment)
- `## C. Verify the configuration` (current Step 4)

With subsections at `### 1.`, `### 2.`, etc.

**Why dismissed:**
- The page mixes a setup procedure with reference sections (`## Auto-provisioned roles`, `## Service accounts`, `## Troubleshooting`, `## FAQ`) — a lettered hierarchy across the whole page would feel forced.
- `rule_007`'s `do_not_apply_when` is borderline-applicable ("single phase / no meaningful grouping").
- Same `heading_hierarchy` constraint as above.

Surfaced in case you want to lean into a lettered structure for the procedural half — but the current `## Step N.` form is also fine.

Create an OIDC application in your identity provider and note the **issuer URL**
and **client ID**. You will need these to configure Materialize.

Set the redirect URI to:

```
https://<your-console-domain>/auth/callback
```

Replace `<your-console-domain>` with the domain where your Materialize console
is accessible.

{{< tabs >}}
{{< tab "Okta" >}}

> The following steps create the OIDC application for the **Materialize console**

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5. `Materialize Console` is the canonical capitalization for the product name.

```suggestion
The following steps create the OIDC application for the **Materialize Console**
```

(browser-based login). If you also need service accounts, you will create
additional applications in the [Service accounts](#service-accounts) section.

1. In the Okta Admin Console, go to **Applications** > **Applications** and
   click **Create App Integration**.

1. Select **OIDC - OpenID Connect** as the sign-in method and **Single-Page
   Application** as the application type. Click **Next**.

1. Configure the application:
   - **App integration name**: Enter a name (e.g., `Materialize`).
   - **Grant type**: Ensure **Authorization Code** is selected (PKCE is used
     automatically for single-page applications).
   - **Sign-in redirect URIs**: Enter
     `https://<your-console-domain>/auth/callback`.
   - **Sign-out redirect URIs**: Optionally, enter
     `https://<your-console-domain>/account/login`.

1. Click **Save**.

1. On the application's **General** tab, note the **Client ID**.

>    *Single-page applications use PKCE instead of a client secret. You do not
>    need a client secret for the console application.*

Reviewer 3: [`rule_015`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_015.yaml) — *reformat inline italic asides as callout blocks*, confidence 2 (**optional**, judgmental). The corpus signal is fuzzy here — italic asides may be a deliberate light-weight style. Feel free to leave as-is.

```suggestion
   {{< note >}}

   Single-page applications use PKCE instead of a client secret. You do not
   need a client secret for the console application.

   {{</ note >}}
```

1. Go to **Security** > **API** and note your **Issuer URI** from the
   authorization server you want to use (e.g.,
   `https://your-org.okta.com/oauth2/default`).

1. Go to the **Assignments** tab and assign the users or groups that should have
   access to Materialize.

>    When a user authenticates via SSO, Materialize uses a JWT claim to determine

Reviewer 3: **Editorial observation (no rule fired):** This exact paragraph appears verbatim three times — at L84–85, L154–155, and L174–175, once per IdP tab (Okta, Entra ID, Generic OIDC). Worth considering moving it above the `{{< tabs >}}` block so it's stated once, since the content is IdP-agnostic. Not a one-click fix, hence no suggestion block.

      1. Under **Assigned to**, select the clients (applications) this rule
         applies to.
      1. Click **Create Rule**.

{{< /tab >}}
{{< tab "Microsoft Entra ID" >}}

1. In the [Azure portal](https://portal.azure.com), go to **Microsoft Entra
   ID** > **App registrations** and click **New registration**.

1. Configure the registration:
   - **Name**: Enter a name (e.g., `Materialize`).
   - **Supported account types**: Select the appropriate option for your
     organization (typically **Accounts in this organizational directory only**).
   - **Redirect URI**: Select **Single-page application (SPA)** and enter
     `https://<your-console-domain>/auth/callback`.

1. Click **Register**.

1. If you plan to use the [Resource Owner Password
   flow](#resource-owner-password-flow) for service accounts, go to
   **Authentication** and set **Allow public client flows** to **Yes**.

1. On the application's **Overview** page, note the **Application (client) ID**
   and the **Directory (tenant) ID**.

1. Go to **Certificates & secrets** > **New client secret**. Add a description
   and expiration, then click **Add**. Note the secret **Value**.

>    *A client secret is not required for the console login (which uses
>    authorization code with PKCE), but is needed if you plan to use the
>    [Client Credentials flow](#client-credentials-flow) for service accounts.*

Reviewer 3: [`rule_015`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_015.yaml) — *reformat inline italic asides as callout blocks*, confidence 2 (**optional**, judgmental).

```suggestion
   {{< note >}}

   A client secret is not required for the console login (which uses
   authorization code with PKCE), but is needed if you plan to use the
   [Client Credentials flow](#client-credentials-flow) for service accounts.

   {{</ note >}}
```

The following example Kubernetes manifest includes configuration for OIDC
authentication:

```hc {hl_lines="15 25"}
apiVersion: v1
kind: Namespace
metadata:
  name: materialize-environment
---
apiVersion: v1
kind: Secret
metadata:
  name: materialize-backend
  namespace: materialize-environment
stringData:
  metadata_backend_url: "..."
  persist_backend_url: "..."
  license_key: "..."
  external_login_password_mz_system: "enter_mz_system_password"
---
apiVersion: materialize.cloud/v1alpha1
kind: Materialize
metadata:
  name: 12345678-1234-1234-1234-123456789012
  namespace: materialize-environment
spec:
  environmentdImageRef: materialize/environmentd:v26.18.0 # Use v26.18.0 or later
  backendSecretName: materialize-backend
  authenticatorKind: Oidc
  requestRollout: 00000000-0000-0000-0000-000000000003 # Switching to Oidc requires a rollout
```

> Apply the updated manifest to your Kubernetes cluster. See

Reviewer 2: So ... from this, I gather you intend this page to be used by people going from None -> Oidc?

Reviewer 2: But why would apply at this step?  Why wouldn't I/couldn't I add the Configmap with the OIDC params?

Author: > But why would apply at this step? Why wouldn't I/couldn't I add the Configmap with the OIDC params?

I think because they could also not set it up with a configmap. So the control flow looks like
```
Enable OIDC authentication (yaml example) -> Configure OIDC system parameters 
                                                   |-> Configure via ConfigMap (yaml example)
                                                   |-> Configure via SQL
```

Reviewer 2: 👍 although ... eek if people use SQL to configure

[Upgrading](/self-managed-deployments/upgrading/#rollout-configuration) for

> details on rollout configuration.

Reviewer 2: Did you intend for people to rollout here? It's not exactly usable at this point.

Author: The intent was to have people rollout here! It not being usable was a bug given it should be backwards compatible with password auth

| `oidc_audience` | A JSON array of expected audience values for token validation (e.g., `["your-client-id"]`). Use the **client ID from [Step 1](#step-1-configure-your-identity-provider)**. Materialize checks that the JWT's `aud` claim contains at least one of these values. **By default, this is empty, and audience validation is skipped.**| No | `[]` |
| `oidc_authentication_claim` | The JWT claim to use as the Materialize username. | No | `sub` |
| `console_oidc_client_id` | The OIDC client ID used by the web console for the authorization code flow. | For console login | Empty |
| `console_oidc_scopes` | Space-separated OIDC scopes requested by the web console when obtaining a token. Scopes control which claims are included in the token. The `openid` scope is required to obtain an ID token. Add `email` to include the `email` claim, or `profile` to include name claims. If `oidc_authentication_claim` references a claim like `email`, you must request the corresponding scope here. | For console login | Empty |

> {{< warning >}}
> When `oidc_audience` is empty, audience validation is skipped. This means
> **any** valid token from the same identity provider can authenticate to
> Materialize, including tokens issued for other applications. **Always set
> `oidc_audience` in production environments.**
> {{</ warning >}}

Reviewer 3: [`rule_043`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_043.yaml) — *normalize blank-line padding inside admonition blocks*, confidence 5.

```suggestion
{{< warning >}}

When `oidc_audience` is empty, audience validation is skipped. This means
**any** valid token from the same identity provider can authenticate to
Materialize, including tokens issued for other applications. **Always set
`oidc_audience` in production environments.**

{{</ warning >}}
```

> ### Mapping IdP Users to Materialize Roles

Reviewer 3: **Editorial consistency (no rule fired):** Title Case heading is inconsistent with peer H3 headings on this page (e.g., `### Configure via ConfigMap`, `### Configure via SQL`, `### Pre-provisioning roles`). House style is sentence case.

```suggestion
### Mapping IdP users to Materialize roles
```

> ### Mapping IdP Users to Materialize Roles

Reviewer 2: How does this Mapping relate to the Configure OIDC system parameters?

Author: Ah it's because I reference `oidc_authentication_claim` right below which is a system parameter! i.e. I'm telling the user "hey this is what this system param does. Make sure to correctly choose it"

When a user authenticates into Materialize, their role name is the value of the JWT claim keyed by `oidc_authentication_claim`.

For example, if `oidc_authentication_claim` is set to `email` and a user authenticates with the following JWT:

```json
{
  "sub": "auth0|abc123",
  "email": "alice@your-org.com",
  "name": "Alice",
  "iat": 1516239022
}
```

Their role name will be `alice@your-org.com`.

For ID tokens (human users), a common claim is `email`. For access tokens from the [Client Credentials flow](#client-credentials-flow), ensure this claim exists in the token.

> If a user tries to login and the role doesn't exist, the role will be auto-provisioned. See [Auto-provisioned roles](#auto-provisioned-roles) for more details.

Reviewer 3: **Author bug (no rule fired):** `login` is a noun; `log in` is the verb. House style uses the verb form here.

```suggestion
If a user tries to log in and the role doesn't exist, the role will be auto-provisioned. See [Auto-provisioned roles](#auto-provisioned-roles) for more details.
```

### Configure via ConfigMap

Create a ConfigMap with your OIDC parameters and reference it in the Materialize
CR's `spec.systemParameterConfigmapName` field. At this point, your manifest should look like:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: mz-system-params
  namespace: materialize-environment
data:
  system-params.json: |
    {
      "oidc_issuer": "YOUR_OIDC_ISSUER",
      "oidc_audience": "[\"YOUR_CLIENT_ID\"]",
      "oidc_authentication_claim": "email",
      "console_oidc_client_id": "YOUR_CLIENT_ID",
      "console_oidc_scopes": "openid email"
    }
---
apiVersion: v1
kind: Secret
metadata:
  name: materialize-backend
  namespace: materialize-environment
stringData:
  metadata_backend_url: ...
  persist_backend_url: ...
  license_key: ...
  external_login_password_mz_system: "enter_mz_system_password"
---
apiVersion: materialize.cloud/v1alpha1
kind: Materialize
metadata:
  name: 12345678-1234-1234-1234-123456789012
  namespace: materialize-environment
spec:
  authenticatorKind: Oidc
  backendSecretName: materialize-backend
  requestRollout: 00000000-0000-0000-0000-000000000003 # Switching to Oidc requires a rollout

>   systemParameterConfigmapName: mz-system-params

Reviewer 2: This kind of strands people.  
You stated above to create a configMap. Did you want people to do something with this?

Author: It's more of an example to show what the configmap should look like, but I don't think it strands people given if they set the values as the placeholders, then nothing should break!

Reviewer 2: So ... I meant above, you had people apply the changes whereas here you don't.  So ... either we don't mention apply in both or do mention in both.  I would link to /self-managed-deployments/configuration-system-parameters/ so that people know what to do.

Author: Gotchya. Makes sense to me!

```

### Configure via SQL

Alternatively, connect as `mz_system` and set the parameters using
`ALTER SYSTEM SET`. The `mz_system` user always authenticates with a password,
even when OIDC is enabled.

```mzsql
ALTER SYSTEM SET oidc_issuer = 'https://your-org.okta.com/oauth2/default';
ALTER SYSTEM SET oidc_audience = '["YOUR_CLIENT_ID"]';
ALTER SYSTEM SET oidc_authentication_claim = 'email';
ALTER SYSTEM SET console_oidc_client_id = 'YOUR_CLIENT_ID';
ALTER SYSTEM SET console_oidc_scopes = 'openid email';
```

## Step 4. Verify the configuration

> 1. Navigate to your Materialize console. You should see an option to sign in

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5.

```suggestion
1. Navigate to your Materialize Console. You should see an option to sign in
```

> 1. Navigate to your Materialize console. You should see an option to sign in

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5.

```suggestion
1. Navigate to your Materialize Console. You should see an option to sign in
```

   via your identity provider via "Use single sign-on".

   ![Materialize console login screen showing the SSO sign-in
   option](/images/console/console-self-managed-sso.png "Materialize console login screen

>    with SSO option")

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5. Image alt text and title attribute (both are user-visible — screen readers + tooltips).

```suggestion
   ![Materialize Console login screen showing the SSO sign-in
   option](/images/console/console-self-managed-sso.png "Materialize Console login screen
   with SSO option")
```

1. Sign in through your IdP. After successful authentication, you are redirected

>    back to the Materialize console.

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5.

```suggestion
   back to the Materialize Console.
```

> 1. To confirm which role you've signed in as via SSO, open the [SQL Shell](/console/sql-shell/) in the Materialize console. In the welcome message, you should see the role name labelled under "User". This is derived from the `oidc_authentication_claim` claim in your identity token:

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5. Plus a small US-English fix: `labelled` → `labeled` (no rule, just house spelling).

```suggestion
1. To confirm which role you've signed in as via SSO, open the [SQL Shell](/console/sql-shell/) in the Materialize Console. In the welcome message, you should see the role name labeled under "User". This is derived from the `oidc_authentication_claim` claim in your identity token:
```

> 1. To confirm which role you've signed in as via SSO, open the [SQL Shell](/console/sql-shell/) in the Materialize console. In the welcome message, you should see the role name labelled under "User". This is derived from the `oidc_authentication_claim` claim in your identity token:

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5. Plus a small US-English fix: `labelled` → `labeled` (no rule, just house spelling).

```suggestion
1. To confirm which role you've signed in as via SSO, open the [SQL Shell](/console/sql-shell/) in the Materialize Console. In the welcome message, you should see the role name labeled under "User". This is derived from the `oidc_authentication_claim` claim in your identity token:
```

> ![Materialize console Shell](/images/console/console.png "Materialize console Shell")

Reviewer 3: [`rule_022`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_022.yaml) — *capitalize Materialize product edition and deployment names*, confidence 5. Image alt text and title.

```suggestion
![Materialize Console Shell](/images/console/console.png "Materialize Console Shell")
```

## Connecting via SQL clients

To connect to Materialize using a SQL client like `psql`, you need an OIDC ID token and the `oidc_auth_enabled=true` connection option.

{{< important >}}
The `oidc_auth_enabled=true` connection option is **required** for OIDC
authentication over the PostgreSQL wire protocol. Without it, Materialize
defaults to password authentication.

> {{</ important >}}

Reviewer 3: [`rule_043`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_043.yaml) — *normalize blank-line padding inside admonition blocks*, confidence 5.

```suggestion
{{< important >}}

The `oidc_auth_enabled=true` connection option is **required** for OIDC
authentication over the PostgreSQL wire protocol. Without it, Materialize
defaults to password authentication.

{{</ important >}}
```

> <!-- TODO (SangJunBak): Uncomment once v26.19 is released.

Reviewer 3: [`rule_032`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_032.yaml) — *remove inline TODO comments from markdown*, confidence 5. The intent (no developer-facing TODOs in published docs) fires here, but the rule's trigger expects markdown `[//]: # "TODO ..."` form, not HTML comment form. **Question for you:** would you rather remove this block entirely now (and re-add when v26.19 ships), or leave it as-is? The author's name in the comment is the only practical concern. Affects lines 358–371.

### Get your token from the console

The easiest way to get an ID token is through the Materialize console:

1. Sign in to the console via SSO.
1. Click **Connect** in the navigation bar. The connection dialog displays your
   ID token along with connection instructions.
1. Copy the provided ID token.

![Materialize console Connect dialog showing the ID
token](/images/sso-connect-token.png "Materialize console Connect dialog
with ID token")
-->

### Connect with psql

Use the ID token as your password:

```shell
PGOPTIONS='-c oidc_auth_enabled=true' \
PGPASSWORD="<your-id-token>" \
psql -h <materialize-host> -p 6875 -U <username> materialize
```

Replace `<username>` with the value of the authentication claim in your JWT
(e.g., your email address if `oidc_authentication_claim` is set to `email`).

> {{< note >}}
> Materialize validates the token at **connection time only**. Once a connection
> is established, it persists until disconnected, regardless of token expiry.
> {{</ note >}}

Reviewer 3: [`rule_043`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_043.yaml) — *normalize blank-line padding inside admonition blocks*, confidence 5.

```suggestion
{{< note >}}

Materialize validates the token at **connection time only**. Once a connection
is established, it persists until disconnected, regardless of token expiry.

{{</ note >}}
```

### Get a token using CLI tools

You can also obtain an ID token outside the console using OIDC CLI tools such as
[`oidc-agent`](https://indigo-dc.gitbook.io/oidc-agent) or
[`oauth2c`](https://github.com/cloudentity/oauth2c).

## Auto-provisioned roles

Each user who authenticates via OIDC is associated with a single database role
in Materialize through `oidc_authentication_claim`. When a user signs in and no matching role exists, Materialize
**automatically creates** a role for them.

Auto-provisioned roles:
- Have default privileges only.
- Must be granted additional privileges through
  [RBAC](/security/self-managed/access-control/manage-roles/).
- Are not automatically removed when the user is removed from the IdP. See
  [De-provisioning users](#de-provisioning-users) for cleanup instructions.

### Pre-provisioning roles

An administrator can create roles before users login, rather than rely on
auto-provisioning. To pre-provision a role, connect as a superuser and create
the role with a name matching the expected JWT claim value:
```mzsql
CREATE ROLE "alice@your-org.com" WITH LOGIN;
```
Learn how to manage pre-provisioned roles in

> To configure privileges for pre-provisioned roles, see [Manage database roles](/security/self-managed/access-control/manage-roles/).

Reviewer 3: **Author bug (correctness, no rule fired):** Line 418 ends mid-air — "Learn how to manage pre-provisioned roles in" has no object. Looks like an authoring artifact. The next line already covers the intent; suggestion below replaces both lines with just the kept sentence.

```suggestion
To configure privileges for pre-provisioned roles, see [Manage database roles](/security/self-managed/access-control/manage-roles/).
```

via OIDC tokens. There are two approaches depending on what your identity
provider supports.

### Resource Owner Password flow

Use this approach when you need a service account that authenticates with a
username and password to obtain an ID token.

{{< tabs >}}
{{< tab "Okta" >}}

1. In the Okta Admin Console, go to **Applications** > **Applications** and
   click **Create App Integration**.

1. Select **OIDC - OpenID Connect** as the sign-in method and **Native
   Application** as the application type. Click **Next**.

1. Configure the application:
   - **App integration name**: Enter a name (e.g., `Materialize ROPC`).
   - **Grant type**: Enable **Resource Owner Password**.

1. Click **Save** and note the **Client ID** and **Client Secret**.

1. Go to the **Assignments** tab and assign the service account user.

1. Ensure your authorization server's access policy includes a rule that
   allows the **Resource Owner Password** grant type for this application.
   See the authorization server setup in
   [Step 1](#step-1-configure-your-identity-provider).

1. In Okta, create a new user to serve as the service account (e.g.,
   `svc-materialize@your-org.com`).

>    *The Resource Owner Password flow does not support MFA in Okta. The service
>    account user must not have MFA enabled.*

Reviewer 3: [`rule_015`](https://github.com/MaterializeInc/mz-skills/blob/main/.agents/skills/mz-doc-edit/rules/rule_015.yaml) — *reformat inline italic asides as callout blocks*, confidence 2 (**optional**, judgmental).

```suggestion
   {{< note >}}

   The Resource Owner Password flow does not support MFA in Okta. The service
   account user must not have MFA enabled.

   {{</ note >}}
```

   To solve this, create a custom claim (e.g., `sql_username`) in your
   authorization server that maps to `user.email` for ID tokens and to a
   configured value for access tokens:

   1. In the Okta Admin Console, go to **Security** > **API** and select your
      authorization server.

   1. Go to the **Claims** tab and click **Add Claim**.

   1. Configure the claim for **ID tokens** (human users):
      - **Name**: `sql_username`
      - **Include in token type**: **ID Token** (always).
      - **Value type**: **Expression**.
      - **Value**: `appuser.email`
      - **Include in**: **Any scope**.

   1. Click **Create**, then click **Add Claim** again.

   1. Configure the claim for **access tokens** (service accounts):
      - **Name**: `sql_username`
      - **Include in token type**: **Access Token** (always).
      - **Value type**: **Expression**.
      - **Value**: `app.sub`
      - **Include in**: **Any scope**.

   1. Click **Create**.

   1. Set the authentication claim in Materialize:

      ```mzsql
      ALTER SYSTEM SET oidc_authentication_claim = 'sql_username';
      ```

   *If you have multiple service accounts using Client Credentials, each needs

>    its own Okta application.

Reviewer 3: **Author bug (correctness, no rule fired):** Unclosed italic. The opening `*` on L646 has no closing `*` — Markdown will render this as italic from here until the next `*` elsewhere in the file.

```suggestion
   *If you have multiple service accounts using Client Credentials, each needs
   its own Okta application.*
```

---
title: "Single sign-on (SSO)"
description: "Configure OIDC-based single sign-on (SSO) for Self-Managed Materialize."
menu:
  main:

>     parent: "security-sm"

Reviewer 2: this is what makes it on the same level as authentication. make it a child of authentication.

    name: "Single sign-on (SSO)"
    identifier: "sso-sm"
    weight: 9
---

{{< public-preview />}}

Single sign-on (SSO) allows users to authenticate to Self-Managed Materialize
using their organization's identity provider (IdP) via
[OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/).
Instead of managing passwords directly in Materialize, users sign in through
their IdP (e.g., Okta, Microsoft Entra ID) and receive a JWT token that
Materialize validates.

{{< note >}}
SSO handles **authentication** only. Permissions within the database are managed
separately using [role-based access control (RBAC)](/security/self-managed/access-control/).
{{</ note >}}

## Before you begin

Make sure you have:

- An OIDC-capable identity provider (e.g., Okta, Microsoft Entra ID, or any
  provider that supports OpenID Connect).
- Admin access to your Kubernetes cluster where Materialize is deployed.

## Step 1. Configure your identity provider

{{< note >}}

> **Audience**: this step is typically performed by your IT or security team.

Reviewer 2: Not sure what purpose this note serves. If you don't have access, you won't be able to do it. And you should know who would perform this in your org.

Author: Ah so I put this here specifically to address some confusion on who needs to get what gathered from our own dogfooding experience([slack thread](https://materializeinc.slack.com/archives/C0A23PK183Z/p1776099639978319)). The idea I wanted to communicate was:
- People who are Okta admins, you guys follow step step 1. And make sure to get the issuer URL and client ID
- The guys who can actually configure Materialize, you guys can follow step 2 and onwards

> **Audience**: this step is typically performed by your IT or security team.

Reviewer 2: Not sure what this is stating.

Author: Addressed here https://github.com/MaterializeInc/materialize/pull/35591/changes/BASE..8555b56b81b25d08b60b074359692781c7af1cc8#r3318388385

The remaining steps (Step 2 onward) are performed by the Materialize

> infrastructure admin. When Step 1 is complete, hand off the following to the

Reviewer 2: This makes it sound like someone from Materialize will need to perform the steps 2+.

Author: We can change it to be more specific that it's the admin of a customer's Materialize instance

Reviewer 2: I just don't think we need to specify who does what.  It's just a lot of writing that guesses at how organizations work and it just clutters the docs up.  For example, why don't we do it for the monitoring docs?

Materialize admin:

- The OIDC **issuer URL**
- The **client ID** for the console application
- (If using service accounts) the client ID, client secret, and expected
  audience for each service-account application
{{</ note >}}

Create an OIDC application in your identity provider and note the **issuer URL**

> and **client ID**. You will need these to configure Materialize.

Reviewer 2: So, we now have lines 48-49 vs lines 42-44. What should people pay attention to?

Author: We can get rid of the "note the **issuer URL**
and **client ID**." here!

> Set the following redirect URIs on the OIDC application:

Reviewer 2: this makes it sound like these are the steps ... but then, what are the specific OIDC provider tabs?

Author: I guess this part is redundant with the tabs. We can get rid of this paragraph up to `you can omit it if no one in your org will use the CLI flow.`.

```
https://<your-console-domain>/auth/callback
http://localhost:9876/callback
```

Replace `<your-console-domain>` with the domain where your Materialize Console
is accessible. The `http://localhost:9876/callback` URI is used by
[`oauth2c`](#get-a-token-using-cli-tools) to fetch ID tokens from the command
line; you can omit it if no one in your org will use the CLI flow.

{{< tabs >}}
{{< tab "Okta" >}}

The following steps create the OIDC application for the **Materialize Console**
(browser-based login). If you also need service accounts, you will create

> additional applications in the [Service accounts](#service-accounts) section.

Reviewer 2: Did the sentence get smooshed? Probably "in the ..." -> is more like See [Service accounts]

        [Client Credentials service account flow](#client-credentials-flow).

      To add or edit a rule:

      1. Click **Add New Access Policy** (or select an existing policy).
      1. Click **Add Rule** within the policy.
      1. Under **Grant type is**, check the grant types you need.
      1. Under **Assigned to**, select the clients (applications) this rule
         applies to.
      1. Click **Create Rule**.

{{< /tab >}}
{{< tab "Microsoft Entra ID" >}}

1. In the [Azure portal](https://portal.azure.com), go to **Microsoft Entra
   ID** > **App registrations** and click **New registration**.

1. Configure the registration:
   - **Name**: Enter a name (e.g., `Materialize`).
   - **Supported account types**: Select the appropriate option for your
     organization (typically **Accounts in this organizational directory only**).
   - **Redirect URI**: Select **Single-page application (SPA)** and enter
     `https://<your-console-domain>/auth/callback`. After registration, you
     can add `http://localhost:9876/callback` under **Authentication** if you
     want to use the [CLI token flow](#get-a-token-using-cli-tools).

1. Click **Register**.

> 1. If you plan to use the [Resource Owner Password

Reviewer 2: I thought we were registering the Console here?

Author: We're registering an application that represents Materialize that the Console, but also other applications can use to get tokens. So you can set this application to use the Resource Owner Password flow,

| `oidc_authentication_claim` | The JWT claim to use as the Materialize username. | No | `sub` |
| `console_oidc_client_id` | The OIDC client ID used by the web console for the authorization code flow. | For console login | Empty |
| `console_oidc_scopes` | Space-separated OIDC scopes requested by the web console when obtaining a token. Scopes control which claims are included in the token. The `openid` scope is required to obtain an ID token. Add `email` to include the `email` claim, or `profile` to include name claims. If `oidc_authentication_claim` references a claim like `email`, you must request the corresponding scope here. | For console login | Empty |

{{< warning >}}
When `oidc_audience` is empty, audience validation is skipped. This means
**any** valid token from the same identity provider can authenticate to
Materialize, including tokens issued for other applications. **Always set
`oidc_audience` in production environments.**
{{</ warning >}}

> ### Mapping IdP users to Materialize roles

Reviewer 2: Why is this section here? It seems like it should be incorporated into the table above for `oidc_authentication_claim`

Author: I wanted this to be a more in-depth explanation and didn't think this could fit in the table. I think we should reference this section into the table

Reviewer 2: So, you see how above you drew with the 2 configures under the COnfigure OIDC system params?
```
Enable OIDC authentication (yaml example) -> Configure OIDC system parameters 
                                                   |-> Configure via ConfigMap (yaml example)
                                                   |-> Configure via SQL
```
 I think you did what most people would expect in this section.

I would move the Mapping Idp users to materialize roles to Provisioning roles.

In the table, include line 282 into the table.

Author: Sounds good to me!

When a user authenticates into Materialize, their role name is the value of the JWT claim keyed by `oidc_authentication_claim`.

For example, if `oidc_authentication_claim` is set to `email` and a user authenticates with the following JWT:

```json
{
  "sub": "auth0|abc123",
  "email": "alice@your-org.com",
  "name": "Alice",
  "iat": 1516239022
}
```

Their role name will be `alice@your-org.com`.

For ID tokens (human users), a common claim is `email`. For access tokens from the [Client Credentials flow](#client-credentials-flow), ensure this claim exists in the token.

If a user tries to log in and the role doesn't exist, the role will be auto-provisioned. See [Auto-provisioned roles](#auto-provisioned-roles) for more details.

### Configure via ConfigMap

Create a ConfigMap with your OIDC parameters and reference it in the Materialize
CR's `spec.systemParameterConfigmapName` field. At this point, your manifest should look like:

> ```yaml

Reviewer 2: if we could highlight the rows that have changed, it'd be easier for people

Author: Sounds good to me!

  namespace: materialize-environment
stringData:
  metadata_backend_url: ...
  persist_backend_url: ...
  license_key: ...
  external_login_password_mz_system: "enter_mz_system_password"
---
apiVersion: materialize.cloud/v1alpha1
kind: Materialize
metadata:
  name: 12345678-1234-1234-1234-123456789012
  namespace: materialize-environment
spec:
  authenticatorKind: Oidc
  backendSecretName: materialize-backend
  requestRollout: 00000000-0000-0000-0000-000000000003 # Switching to Oidc requires a rollout
  systemParameterConfigmapName: mz-system-params
```

### Configure via SQL

Alternatively, connect as `mz_system` and set the parameters using
`ALTER SYSTEM SET`. The `mz_system` user always authenticates with a password,
even when OIDC is enabled.

```mzsql
ALTER SYSTEM SET oidc_issuer = 'https://your-org.okta.com/oauth2/default';
ALTER SYSTEM SET oidc_audience = '["YOUR_CLIENT_ID"]';
ALTER SYSTEM SET oidc_authentication_claim = 'email';
ALTER SYSTEM SET console_oidc_client_id = 'YOUR_CLIENT_ID';
ALTER SYSTEM SET console_oidc_scopes = 'openid email';
```

## Step 4. Verify the configuration

1. Navigate to your Materialize Console. You should see an option to sign in

>    via your identity provider via "Use single sign-on".

Reviewer 2: too many "via"s  ... not sure they're needed.  basically, you see an option to "Use single sign-on"

Author: Sgtm

### Connect with psql

Use the ID token as your password:

```shell
PGPASSWORD="<your-id-token>" \
psql -h <materialize-host> -p 6875 -U <username> materialize
```

Replace `<username>` with the value of the authentication claim in your JWT
(e.g., your email address if `oidc_authentication_claim` is set to `email`).

{{< note >}}
Materialize validates the token at **connection time only**. Once a connection
is established, it persists until disconnected, regardless of token expiry.
{{</ note >}}

## Auto-provisioned roles

Each user who authenticates via OIDC is associated with a single database role
in Materialize through `oidc_authentication_claim`. When a user signs in and no matching role exists, Materialize
**automatically creates** a role for them.

Auto-provisioned roles:
- Have default privileges only.
- Must be granted additional privileges through
  [RBAC](/security/self-managed/access-control/manage-roles/).
- Are not automatically removed when the user is removed from the IdP. See
  [De-provisioning users](#de-provisioning-users) for cleanup instructions.

> ### Pre-provisioning roles

Reviewer 2: This child section is weird to have in a parent section named auto-provisioned roles. As it is currently, just looking at the headers, it's:

- Auto provisioned roles
  - Pre-provisioning roles
  -  Auditing auto-provisioned roles

Author: I can separate `pre-provisioning roles` into another section adjacent to `Auto provisioned roles`

An administrator can create roles before users login, rather than rely on
auto-provisioning. To pre-provision a role, connect as a superuser and create
the role with a name matching the expected JWT claim value:
```mzsql
CREATE ROLE "alice@your-org.com" WITH LOGIN;
```
To configure privileges for pre-provisioned roles, see [Manage database roles](/security/self-managed/access-control/manage-roles/).

### Auditing auto-provisioned roles

To view which roles were auto-provisioned via OIDC, query `mz_audit_events`:

```mzsql
SELECT details
FROM mz_audit_events
WHERE event_type = 'create' AND object_type = 'role' AND details ->> 'auto_provision_source' = 'oidc'
ORDER BY occurred_at DESC;
```

Roles created through OIDC authentication will have `auto_provision_source` set to
`oidc`.

## Service accounts

> For machine-to-machine access, you have three options. Use **SQL password

Reviewer 2: I would bullet the three options and link to the 3 sections.

Author: Sounds good to me

authentication** unless your service account specifically needs to
authenticate against your IdP — in that case, use one of the OAuth flows
([Resource Owner Password](#resource-owner-password-flow) or
[Client Credentials](#client-credentials-flow)).

> ### SQL password authentication (recommended for non-OAuth clients)

Reviewer 2: I think we need a better narrative on why we're putting this here.

Author: Biggest narrative being some clients (i.e. looker) just won't allow oauth flows. We can add something like "required for clients that don't support oauth flows", but don't think we should give any examples.

Author: Also it is an authentication method that's supported in this mode

The simplest way to give a service or application access to Materialize on an
OIDC listener is to create a role with a SQL password.

Use this approach when your client doesn't speak OAuth natively.

1. As a user with the `CREATEROLE` privilege, create the role with a password:

    ```mzsql
    CREATE ROLE "svc-dbt" WITH LOGIN PASSWORD 'a-strong-password';
    ```

1. Grant the privileges this service account needs. See
   [Manage database roles](/security/self-managed/access-control/manage-roles/)
   for the full privilege model.

1. Connect using the password directly:

    ```shell
    PGPASSWORD="a-strong-password" \
    psql -h <materialize-host> -p 6875 -U svc-dbt materialize
    ```

For dbt-specific setup, see [dbt connection profiles](/manage/dbt/get-started/).
For Terraform, see [Terraform: get started](/manage/terraform/get-started/).

> Use the OAuth-based flows below ([Resource Owner Password](#resource-owner-password-flow),

Reviewer 2: ? Is this sentence a straggler? Not sure why this is here at the end after I set up the password auth.

Author: We can get rid of this paragraph

> [Client Credentials](#client-credentials-flow)) when youneed

Reviewer 2: typo "youneed"

Author: Will fix!

1. Configure the application:
   - **App integration name**: Enter a name (e.g., `Materialize ROPC`).
   - **Grant type**: Enable **Resource Owner Password**.

1. Click **Save** and note the **Client ID** and **Client Secret**.

1. Go to the **Assignments** tab and assign the service account user.

1. Ensure your authorization server's access policy includes a rule that
   allows the **Resource Owner Password** grant type for this application.
   See the authorization server setup in
   [Step 1](#step-1-configure-your-identity-provider).

1. In Okta, create a new user to serve as the service account (e.g.,
   `svc-materialize@your-org.com`).

   *The Resource Owner Password flow does not support MFA in Okta. The service
   account user must not have MFA enabled.*

1. Fetch an ID token:

   ```shell
   curl -X POST https://your-org.okta.com/oauth2/default/v1/token \
     -H "Content-Type: application/x-www-form-urlencoded" \
     -H "Accept: application/json" \
     --data-urlencode "grant_type=password" \
     --data-urlencode "username=svc-materialize@your-org.com" \
     --data-urlencode "password=YOUR_SERVICE_ACCOUNT_PASSWORD" \

>      --data-urlencode "scope=openid email" \

Reviewer 2: So ... here we list both `openid` and `email` ... the other two tabs don't.  Is that intentional?

Author: Ah it's not, although to get the `email` claim, you do need the `email` scope. I can omit this however

Author: Actually it seems like our other examples include `openid` and `email`. I'll leave it in for ROPC

     --data-urlencode "client_id=YOUR_ROPC_CLIENT_ID" \
     --data-urlencode "client_secret=YOUR_ROPC_CLIENT_SECRET"
   ```

1. Extract the `id_token` from the JSON response and use it to connect:

   ```shell
   PGPASSWORD="<id-token>" \
   psql -h <materialize-host> -p 6875 -U svc-materialize@your-org.com materialize
   ```

{{< /tab >}}
{{< tab "Microsoft Entra ID" >}}

1. In Microsoft Entra ID, create a new user to serve as the service account.

1. Assign the user to your Materialize application under **Enterprise
   applications** > **Users and groups**.

1. Ensure **Allow public client flows** is set to **Yes** in your app
   registration's **Authentication** settings. If you followed

>    [Step 1](#step-1-configure-your-identity-provider), this is already

Reviewer 2: But ... doesn't Step 1 sets up Materialize Console and not your app, yes?

Author: We technically could re-use the Materialize Console app for this flow, but the other tabs have you create a new application/client so we should just do that instead.

     -H "Content-Type: application/x-www-form-urlencoded" \
     --data-urlencode "grant_type=password" \
     --data-urlencode "username=svc-materialize@your-org.com" \
     --data-urlencode "password=YOUR_SERVICE_ACCOUNT_PASSWORD" \
     --data-urlencode "scope=openid" \
     --data-urlencode "client_id=YOUR_CLIENT_ID"
   ```

1. Extract the `id_token` from the JSON response and use it to connect:

   ```shell
   PGPASSWORD="<id-token>" \
   psql -h <materialize-host> -p 6875 -U svc-materialize@your-org.com materialize
   ```

{{< /tab >}}
{{< tab "Generic OIDC" >}}

1. Create a service account user in your identity provider.

1. Assign the service account to your Materialize application.

1. Enable the Resource Owner Password Credentials grant for your application.

1. Fetch an ID token from your IdP's token endpoint:

   ```shell
   curl -X POST https://your-idp.com/oauth2/token \
     -H "Content-Type: application/x-www-form-urlencoded" \
     --data-urlencode "grant_type=password" \
     --data-urlencode "username=svc-materialize@your-org.com" \
     --data-urlencode "password=YOUR_SERVICE_ACCOUNT_PASSWORD" \

>      --data-urlencode "scope=openid" \

Reviewer 2: ditto.

Author: We can leave this as `openid`

Author: Actually it seems like our other examples include openid and email. I'll leave it in for ROPC

     --data-urlencode "client_id=YOUR_CLIENT_ID" \
     --data-urlencode "client_secret=YOUR_CLIENT_SECRET"
   ```

1. Extract the `id_token` from the JSON response and use it to connect:

   ```shell
   PGPASSWORD="<id-token>" \
   psql -h <materialize-host> -p 6875 -U svc-materialize@your-org.com materialize
   ```

{{< /tab >}}
{{< /tabs >}}

### Client Credentials flow

Use this approach to treat an IdP client as a service account. This is useful
for automated systems that do not have a user context.

The `oidc_authentication_claim` setting determines which token claim maps to the
Materialize role name. Note that Client Credentials tokens are **access tokens**
(not ID tokens) and may not include the same claims as user tokens. In

> particular, the default `sub` claim contains the client ID, not a

Reviewer 2: So .. ? you talk about `sub` claim here ... and below, in some tabs, wee seem to only talk about `aud` claim.  coming in blind so I'm not sure whether we probably should clarify

Author: This is more of a warning than anything. We can just get rid of this paragraph.

human-readable name.

{{< tabs >}}
{{< tab "Okta" >}}

1. In the Okta Admin Console, go to **Applications** > **Applications** and
   click **Create App Integration**.

1. Select **OIDC - OpenID Connect** as the sign-in method and **Web
   Application** as the application type. Click **Next**.

1. Configure the application:
   - **App integration name**: Enter a name (e.g., `Materialize Service Account 1`).
   - **Grant type**: Enable **Client Credentials** (deselect other grant
     types).

1. Click **Save** and note the **Client ID** and **Client Secret**.

1. Ensure your authorization server's access policy includes a rule that
   allows the **Client Credentials** grant type for this application.
   See the authorization server setup in
   [Step 1](#step-1-configure-your-identity-provider).

1. Ensure `oidc_audience` includes the expected audience value for tokens
   from your authorization server. In Okta, the `aud` claim is set to the
   authorization server's audience (configured in **Security** > **API** >
   your auth server > **Settings**), not the client ID. For the default
   authorization server, this is typically `api://default`:

   ```mzsql

>    ALTER SYSTEM SET oidc_audience = '["api://default"]';

Reviewer 2: So ... people will probably copy this .. and wipe out the Console? Why not have it be an array of values? (But, not sure if intentional ...)

Author: You're right. I showed this step moreso for demonstration. We can add a note to check for existing audiences and add this audience to the array

     --data-urlencode "grant_type=client_credentials" \
     --data-urlencode "scope=openid" \
     --data-urlencode "client_id=YOUR_SERVICE_CLIENT_ID" \
     --data-urlencode "client_secret=YOUR_SERVICE_CLIENT_SECRET"
   ```

1. Extract the `access_token` from the JSON response and use it to connect:

   ```shell
   PGPASSWORD="<access-token>" \
   psql -h <materialize-host> -p 6875 -U <service-account-name> materialize
   ```

   The `<service-account-name>` must match the value of the `sql_username`
   claim in the access token (e.g., `svc-my-service`).

{{< /tab >}}
{{< tab "Microsoft Entra ID" >}}

1. In the [Azure portal](https://portal.azure.com), go to **Microsoft Entra
   ID** > **App registrations** and click **New registration**.

1. Configure the registration:
   - **Name**: Enter a name (e.g., `Materialize Service Account`).
   - **Supported account types**: Select the appropriate option for your
     organization.

1. Click **Register**.

1. On the application's **Overview** page, note the **Application (client) ID**.

1. Go to **Certificates & secrets** > **New client secret**. Add a description
   and expiration, then click **Add**. Note the secret **Value**.

> 1. Ensure `oidc_audience` includes the expected audience value for Client

Reviewer 2: Should this step come after the Fetch an access token step?

Author: Good point, it should!

---
title: "Single sign-on (SSO)"
description: "Configure OIDC-based single sign-on (SSO) for Self-Managed Materialize."
menu:
  main:
    parent: "authentication-sm"
    name: "Single sign-on (SSO)"
    identifier: "sso-sm"
    weight: 1
---

{{< public-preview />}}

Single sign-on (SSO) allows users to authenticate to Self-Managed Materialize
using their organization's identity provider (IdP) via
[OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/).
Instead of managing passwords directly in Materialize, users sign in through
their IdP (e.g., Okta, Microsoft Entra ID) and receive a JWT token that
Materialize validates.

{{< note >}}
SSO handles **authentication** only. Permissions within the database are managed
separately using [role-based access control (RBAC)](/security/self-managed/access-control/).
{{</ note >}}

## Before you begin

Make sure you have:

- An OIDC-capable identity provider (e.g., Okta, Microsoft Entra ID, or any
  provider that supports OpenID Connect).
- Admin access to your Kubernetes cluster where Materialize is deployed.

## Step 1. Configure your identity provider

{{< note >}}

> The remaining steps (Step 2 onward) will need:

Reviewer 2: Because of the placement at the top of the steps,  "The remaining steps (Step 2 onward) ..." is confusing 

Let's just go with:
You will use the following values from
your IdP configuration to [configure OIDC system parameters for Materialize](#step-3-configure-oidc-system-parameters):

- ...

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

- The OIDC **issuer URL**
- The **client ID** for the console application

> - (If using service accounts) the client ID, client secret, and expected

Reviewer 2: Why parenthesize the If using service accounts? It's a key component of that bullet point.

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

   the role name. See [Mapping IdP users to Materialize roles](#mapping-idp-users-to-materialize-roles) for more details.

{{< /tab >}}
{{< tab "Generic OIDC" >}}

1. In your identity provider, create a new OIDC **public** client application
   (single-page application type) with the **Authorization Code** grant type
   and **PKCE** support.

1. Set the redirect URI to `https://<your-console-domain>/auth/callback`. If
   you want to use the [CLI token flow](#get-a-token-using-cli-tools), also
   add `http://localhost:9876/callback`.

1. Note the **client ID** and **issuer URL** provided by your identity provider.
   The issuer URL is typically the base URL of your identity provider's OIDC
   discovery endpoint (without `/.well-known/openid-configuration`).

1. Ensure the `openid` scope is available.

1. Assign users or groups that should have access to Materialize.

   When a user authenticates via SSO, Materialize uses a JWT claim to determine
   the role name. See [Mapping IdP users to Materialize roles](#mapping-idp-users-to-materialize-roles) for more details.

{{< /tab >}}
{{< /tabs >}}

{{< note >}}

> **Handoff checklist**: before continuing to Step 2, the Materialize admin

Reviewer 2: Same here.

Once you have configured your IdP, you will need the following values to [configure OIDC system parameters for Materialize](#step-3-configure-oidc-system-parameters):

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

needs the following from whoever configured the IdP:

- The OIDC **issuer URL**
- The **client ID** for the console application

> - (If using service accounts) the client ID, client secret, and expected

Reviewer 2: Same comment about the parenthesis here.

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

{{< warning >}}
When `oidc_audience` is empty, audience validation is skipped. This means
**any** valid token from the same identity provider can authenticate to
Materialize, including tokens issued for other applications. **Always set
`oidc_audience` in production environments.**
{{</ warning >}}

### Mapping IdP users to Materialize roles

When a user authenticates into Materialize, their role name is the value of the JWT claim keyed by `oidc_authentication_claim`.

For example, if `oidc_authentication_claim` is set to `email` and a user authenticates with the following JWT:

```json
{
  "sub": "auth0|abc123",
  "email": "alice@your-org.com",
  "name": "Alice",
  "iat": 1516239022
}
```

Their role name will be `alice@your-org.com`.

For ID tokens (human users), a common claim is `email`. For access tokens from the [Client Credentials flow](#client-credentials-flow), ensure this claim exists in the token.

If a user tries to log in and the role doesn't exist, the role will be auto-provisioned. See [Auto-provisioned roles](#auto-provisioned-roles) for more details.

### Configure via ConfigMap

Create a ConfigMap with your OIDC parameters and reference it in the Materialize

> CR's `spec.systemParameterConfigmapName` field. At this point, your manifest should look like:

Reviewer 2: probably should blurb that in this example, we're using auth claim `email` instead of the default sub 

maybe (?)repeat the blurb about including in the oidc_scopes as well.  It's standard practice w.r.t. scopes, so not quite sure we need to repeat the blurb, so either or.

Author: We can repeat the blurb about including in the oidc_scopes 

>Space-separated OIDC scopes requested by the web console when obtaining a token. Scopes control which claims are included in the token. The `openid` scope is required to obtain an ID token. Add `email` to include the `email` claim, or `profile` to include name claims. If `oidc_authentication_claim` references a claim like `email`, you must request the corresponding scope here.

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

      --grant-type authorization_code \
      --pkce \
      --scopes 'openid email' \
      --auth-method none \
      --silent | jq -r '.id_token'
    ```

    A browser window opens to complete the IdP login. After signing in,
    `oauth2c` prints the ID token to stdout.

ID tokens expire (typically within an hour). Re-run the command above when
your token expires.

### Get a token from the Materialize console

Clicking "Connect" in the Materialize console will provide you with an ID token that you can use to connect.

![Materialize Console connect instructions for OIDC](/images/console/console-connect-oidc.png "Materialize Console connect screen for OIDC")

### Connect with psql

Use the ID token as your password:

```shell
PGPASSWORD="<your-id-token>" \
psql -h <materialize-host> -p 6875 -U <username> materialize
```

Replace `<username>` with the value of the authentication claim in your JWT
(e.g., your email address if `oidc_authentication_claim` is set to `email`).

{{< note >}}
Materialize validates the token at **connection time only**. Once a connection
is established, it persists until disconnected, regardless of token expiry.
{{</ note >}}

> ## Auto-provisioned roles

Reviewer 2: I would create a 
Provision role section
- that Mapping Idp users to materialize roles section from above to here
- Auto-provisioning roles
   - Auditing auto-provisioned roles 
- Pre-provisioning roles

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

in Materialize through `oidc_authentication_claim`. When a user signs in and no matching role exists, Materialize
**automatically creates** a role for them.

Auto-provisioned roles:
- Have default privileges only.
- Must be granted additional privileges through
  [RBAC](/security/self-managed/access-control/manage-roles/).
- Are not automatically removed when the user is removed from the IdP. See
  [De-provisioning users](#de-provisioning-users) for cleanup instructions.

### Auditing auto-provisioned roles

To view which roles were auto-provisioned via OIDC, query `mz_audit_events`:

```mzsql
SELECT details
FROM mz_audit_events
WHERE event_type = 'create' AND object_type = 'role' AND details ->> 'auto_provision_source' = 'oidc'
ORDER BY occurred_at DESC;
```

Roles created through OIDC authentication will have `auto_provision_source` set to
`oidc`.

## Pre-provisioning roles

An administrator can create roles before users login, rather than rely on
auto-provisioning. To pre-provision a role, connect as a superuser and create
the role with a name matching the expected JWT claim value:

```mzsql
CREATE ROLE "alice@your-org.com" WITH LOGIN;
```

> To configure privileges for pre-provisioned roles, see [Manage database roles](/security/self-managed/access-control/manage-roles/).

Reviewer 2: how come we don't mention privileges for auto provisioned roles?

Author: We do note below:
> Auto-provisioned roles:  
	- Have default privileges only. 

but I can extend this to both pre-provisioned and auto-provisioned roles!

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

## Service accounts

For machine-to-machine access, you have three options:

- [SQL password authentication](#sql-password-authentication-recommended-for-non-oauth-clients):
  for clients that don't support OAuth flows.
- [Resource Owner Password flow](#resource-owner-password-flow): for service
  accounts that authenticate against your IdP with a username and password.
- [Client Credentials flow](#client-credentials-flow): for service accounts
  that authenticate against your IdP without a user context.

> Use SQL password authentication unless your service account specifically needs

Reviewer 2: So ... we state in the next line that password auth is recommended for non-oauth clients ...
whereas ... this sentence recommends password auth in general.   Which is it?

Author: Ah this is a straggler. I think we should get rid of this sentence.

Author: Addressed here [`fb4471d` (this PR)](https://github.com/MaterializeInc/materialize/pull/35591/changes/fb4471d20364b064ddd21e352535f684db4beb02)

   `email`, human users get readable role names (e.g., `alice@your-org.com`),
   but Client Credentials access tokens do not include an `email` claim by
   default. If set to `sub`, Client Credentials tokens work, but human users
   lose email-based role names and get opaque subject IDs instead.

   To solve this, create a custom claim (e.g., `sql_username`) in your
   authorization server that maps to `user.email` for ID tokens and to a
   configured value for access tokens:

   1. In the Okta Admin Console, go to **Security** > **API** and select your
      authorization server.

   1. Go to the **Claims** tab and click **Add Claim**.

   1. Configure the claim for **ID tokens** (human users):
      - **Name**: `sql_username`
      - **Include in token type**: **ID Token** (always).
      - **Value type**: **Expression**.
      - **Value**: `appuser.email`
      - **Include in**: **Any scope**.

   1. Click **Create**, then click **Add Claim** again.

   1. Configure the claim for **access tokens** (service accounts):
      - **Name**: `sql_username`
      - **Include in token type**: **Access Token** (always).
      - **Value type**: **Expression**.
      - **Value**: `app.sub`

>       - **Include in**: **Any scope**.

Reviewer 2: ? does the scope need to change for custom claims?

Author: Not exactly! So this just means "what scopes need to be requested for this to be included". It's upto the admin/user to set this up but for simplicity sake, we just say "hey include this in any scope". 

<img width="701" height="154" alt="Image" src="https://github.com/user-attachments/assets/fe2d6e06-f0fd-4226-9234-958a96f59315" />
